### ISSUE TITLE
Canceled commands lead to connection desynchronization

### ISSUE THREAD

**Issue by long-xuan-nguyen** [mp911de/lettuce#262]
This is a continuation of the issue https://github.com/mp911de/lettuce/issues/253

I've applied the workaround, which is to disable pingBeforeActivate. Last night, a networking event occurred and one of the redis host connection got cut off for a couple of minutes and i am noticing some new issues.

Setup: 
- Redis 3.0.7
- 20 master nodes, each have 2 slave nodes
- Lettuce 4.1
- ClusterClientOption:
    .closeStaleConnections(true)
    .refreshClusterView(true)
    .refreshPeriod(60, TimeUnit.SECONDS)
    .validateClusterNodeMembership(true)
    .pingBeforeActivateConnection(false)
    .autoReconnect(true)
    .build

What happened:

- Slave S1 network got cut off
- after a while, got RedisCommandTimeoutException, which is fine
- Some of the commands got written to the wire
- localhost:port1 -> S1 connection got closed
- These commands got written to S2 now
- CommandOutput is not populated, leaving the output as null. All subsequent requests to S2 now render the output null
- A new connection (new local port) to S1 is made and it works OK. 

Here are some logs:

```
03 Jun 2016 02:13:18,142 [DEBUG] fa1beb32-ce27-42f9-b9f7-811bcc7c9c34 (http-bio-0.0.0.0-1080-exec-113) *****: Processing CountQueryRequest(arg=aaa)
03 Jun 2016 02:13:18,142 [DEBUG] fa1beb32-ce27-42f9-b9f7-811bcc7c9c34 (http-bio-0.0.0.0-1080-exec-113) com.lambdaworks.redis.RedisChannelHandler: dispatching command AsyncCommand [type=ZCARD, commandType=com.lambdaworks.redis.protocol.Command] with args aaa
03 Jun 2016 02:13:18,142 [DEBUG] fa1beb32-ce27-42f9-b9f7-811bcc7c9c34 (http-bio-0.0.0.0-1080-exec-113) com.lambdaworks.redis.cluster.PooledClusterConnectionProvider: getConnection(READ, 8141)
03 Jun 2016 02:13:18,142 [DEBUG] fa1beb32-ce27-42f9-b9f7-811bcc7c9c34 (http-bio-0.0.0.0-1080-exec-113) com.lambdaworks.redis.cluster.PooledClusterConnectionProvider: returned connection is com.lambdaworks.redis.StatefulRedisConnectionImpl@4392cdd1
03 Jun 2016 02:13:18,142 [DEBUG] fa1beb32-ce27-42f9-b9f7-811bcc7c9c34 (http-bio-0.0.0.0-1080-exec-113) com.lambdaworks.redis.protocol.CommandHandler: [/10.81.143.248:39846 -> /10.88.161.176:6379] write() writeAndFlush Command ClusterCommand [command=AsyncCommand [type=ZCARD, commandType=com.lambdaworks.redis.protocol.Command], executions=0, arg=0:0/3/15636812025
03 Jun 2016 02:13:18,142 [DEBUG] fa1beb32-ce27-42f9-b9f7-811bcc7c9c34 (http-bio-0.0.0.0-1080-exec-113) com.lambdaworks.redis.protocol.CommandHandler: [/10.81.143.248:39846 -> /10.88.161.176:6379] Command wrtten: ClusterCommand [command=AsyncCommand [type=ZCARD, commandType=com.lambdaworks.redis.protocol.Command], executions=0, arg=0:0/3/15636812025
03 Jun 2016 02:13:18,142 [DEBUG] fa1beb32-ce27-42f9-b9f7-811bcc7c9c34 (http-bio-0.0.0.0-1080-exec-113) com.lambdaworks.redis.protocol.CommandHandler: [/10.81.143.248:39846 -> /10.88.161.176:6379] write() done
03 Jun 2016 02:13:18,142 [DEBUG] fa1beb32-ce27-42f9-b9f7-811bcc7c9c34 (http-bio-0.0.0.0-1080-exec-113) com.lambdaworks.redis.AbstractRedisAsyncCommands: Dispatched to the underlying connection Command [type=ZCARD, output=IntegerOutput [output=null, error='null']]
03 Jun 2016 02:13:33,206 [DEBUG]  (lettuce-eventExecutorLoop-1-3) com.lambdaworks.redis.protocol.CommandHandler: [/10.81.143.248:45146 -> /10.80.26.133:6379] write() writeAndFlush Command ClusterCommand [command=AsyncCommand [type=ZCARD, commandType=com.lambdaworks.redis.protocol.Command], executions=0, arg=0:0/3/15636812025
03 Jun 2016 02:13:33,206 [DEBUG]  (lettuce-eventExecutorLoop-1-3) com.lambdaworks.redis.protocol.CommandHandler: [/10.81.143.248:45146 -> /10.80.26.133:6379] Command wrtten: ClusterCommand [command=AsyncCommand [type=ZCARD, commandType=com.lambdaworks.redis.protocol.Command], executions=0, arg=0:0/3/15636812025
03 Jun 2016 02:13:33,209 [DEBUG] 1d349f58-552e-4fc8-b844-01af5a04b734 (lettuce-nioEventLoop-3-144) com.lambdaworks.redis.protocol.CommandHandler: [/10.81.143.248:45146 -> /10.80.26.133:6379] Queue contains: 13 commands. Processing ClusterCommand [command=AsyncCommand [type=ZCARD, commandType=com.lambdaworks.redis.protocol.Command], executions=0, arg=0:0/3/15636812025
03 Jun 2016 02:13:33,209 [DEBUG] 1d349f58-552e-4fc8-b844-01af5a04b734 (lettuce-nioEventLoop-3-144) com.lambdaworks.redis.protocol.RedisStateMachine: Decode ClusterCommand [command=AsyncCommand [type=ZCARD, commandType=com.lambdaworks.redis.protocol.Command], executions=0, arg=0:0/3/15636812025
03 Jun 2016 02:13:33,209 [DEBUG] 1d349f58-552e-4fc8-b844-01af5a04b734 (lettuce-nioEventLoop-3-144) com.lambdaworks.redis.protocol.RedisStateMachine: Decoded ClusterCommand [command=AsyncCommand [type=ZCARD, commandType=com.lambdaworks.redis.protocol.Command], executions=0, arg=0:0/3/15636812025, empty stack: true
03 Jun 2016 02:13:33,209 [DEBUG] 1d349f58-552e-4fc8-b844-01af5a04b734 (lettuce-nioEventLoop-3-144) com.lambdaworks.redis.protocol.CommandHandler: [/10.81.143.248:45146 -> /10.80.26.133:6379] Received response for ClusterCommand [command=AsyncCommand [type=ZCARD, commandType=com.lambdaworks.redis.protocol.Command], executions=1, arg=aaa
```

Note how the command is rewritten to 10.80.26.133, which is S2. S1 is 10.88.161.17.

Log showing connection to S1 got closed

```
03 Jun 2016 02:13:27,652 [DEBUG] 9a186dd6-060c-4677-971b-ee71443f92d8 (http-bio-0.0.0.0-1080-exec-101) com.lambdaworks.redis.protocol.CommandHandler: [/10.81.143.248:39846 -> /10.88.161.176:6379] Command wrtten: ClusterCommand [command=AsyncCommand [type=HGET, commandType=com.lambdaworks.redis.protocol.Command], executions=0, arg=0:3029:d
03 Jun 2016 02:13:27,652 [DEBUG] 9a186dd6-060c-4677-971b-ee71443f92d8 (http-bio-0.0.0.0-1080-exec-101) com.lambdaworks.redis.protocol.CommandHandler: [/10.81.143.248:39846 -> /10.88.161.176:6379] write() done
03 Jun 2016 02:13:33,192 [DEBUG]  (lettuce-eventExecutorLoop-1-3) com.lambdaworks.redis.protocol.CommandHandler: [/10.81.143.248:39846 -> /10.88.161.176:6379] close()
03 Jun 2016 02:13:33,192 [DEBUG] 1d349f58-552e-4fc8-b844-01af5a04b734 (lettuce-nioEventLoop-3-142) com.lambdaworks.redis.protocol.ConnectionWatchdog: [unknown] userEventTriggered(ChannelHandlerContext(ConnectionWatchdog#0, [id: 0xa9b16301, L:/10.81.143.248:39846 - R:/10.88.161.176:6379]), com.lambdaworks.redis.ConnectionEvents$PrepareClose@60717ec)
03 Jun 2016 02:13:33,192 [DEBUG] 1d349f58-552e-4fc8-b844-01af5a04b734 (lettuce-nioEventLoop-3-142) com.lambdaworks.redis.protocol.ConnectionWatchdog: [unknown] userEventTriggered(ChannelHandlerContext(ConnectionWatchdog#0, [id: 0xa9b16301, L:/10.81.143.248:39846 ! R:/10.88.161.176:6379]), com.lambdaworks.redis.ConnectionEvents$Close@7581106d)
03 Jun 2016 02:13:33,192 [DEBUG] 1d349f58-552e-4fc8-b844-01af5a04b734 (lettuce-nioEventLoop-3-142) com.lambdaworks.redis.protocol.ConnectionWatchdog: [unknown] channelInactive(ChannelHandlerContext(ConnectionWatchdog#0, [id: 0xa9b16301, L:/10.81.143.248:39846 ! R:/10.88.161.176:6379]))
03 Jun 2016 02:13:33,192 [DEBUG] 1d349f58-552e-4fc8-b844-01af5a04b734 (lettuce-nioEventLoop-3-142) com.lambdaworks.redis.protocol.CommandHandler: [/10.81.143.248:39846 -> /10.88.161.176:6379] channelInactive()
03 Jun 2016 02:13:33,192 [DEBUG] 1d349f58-552e-4fc8-b844-01af5a04b734 (lettuce-nioEventLoop-3-142) com.lambdaworks.redis.protocol.CommandHandler: [/10.81.143.248:39846 -> /10.88.161.176:6379] channel is closed
03 Jun 2016 02:13:33,192 [DEBUG] 1d349f58-552e-4fc8-b844-01af5a04b734 (lettuce-nioEventLoop-3-142) com.lambdaworks.redis.protocol.CommandHandler: [/10.81.143.248:39846 -> /10.88.161.176:6379] deactivating channel handler
03 Jun 2016 02:13:33,192 [DEBUG] 1d349f58-552e-4fc8-b844-01af5a04b734 (lettuce-nioEventLoop-3-142) com.lambdaworks.redis.protocol.CommandHandler: [/10.81.143.248:39846 -> /10.88.161.176:6379] channel is closed
03 Jun 2016 02:13:33,192 [DEBUG] 1d349f58-552e-4fc8-b844-01af5a04b734 (lettuce-nioEventLoop-3-142) com.lambdaworks.redis.protocol.CommandHandler: [/10.81.143.248:39846 -> /10.88.161.176:6379] channel is closed
03 Jun 2016 02:13:33,192 [DEBUG] 1d349f58-552e-4fc8-b844-01af5a04b734 (lettuce-nioEventLoop-3-142) com.lambdaworks.redis.protocol.CommandHandler: [/10.81.143.248:39846 -> /10.88.161.176:6379] channelInactive() done
03 Jun 2016 02:13:33,207 [DEBUG]  (lettuce-eventExecutorLoop-1-3) com.lambdaworks.redis.protocol.CommandHandler: [/10.81.143.248:39846 -> /10.88.161.176:6379] close()
```

I have not been able to reproduce this locally but i thought sending this out would be helpful while i work on reproducing this bug

Make sure that:

- [ ] You have read the [contribution guidelines](https://github.com/mp911de/lettuce/blob/master/.github/CONTRIBUTING.md).
- [ ] You specify the lettuce version and environment so it's obvious which version is affected
- [ ] You provide a reproducible test case (either descriptive of as JUnit test) if it's a bug or the expected behavior differs from the actual behavior.

**Comment by mp911de** [comment]
Thanks for the ticket. I think I can tell you a bit what happened here (at least for certain parts):

1. A topology refresh must have run before closing the S1 connection (don't see that in the logs, hence I'm assuming the refresh)
2. S1 was removed from the topology view, either because it was reconfigured during runtime or `CLUSTER NODES` reported the slave with `NOADDR` (don't see it either in the log)
3. `closeStaleConnections` is enabled and so a topology update will close connections to nodes which are no longer reported as part of the cluster
4. Some internals: Lettuce maintains a set of cluster node connections. Each connection is either identified by `nodeId` or host/port. Connections using `nodeId` can reconnect to the same logical cluster node as the address is obtained from the topology view identified by `nodeId`. Host/port connections have no way to recover themselves. Lettuce uses host/port-identified connections to dispatch its commands.
5. As the connection was closed, all commands get re-dispatched using the most recent topology view and should get executed.

I can't see from your logs the command response is `null`, guess need to investigate that issue.

**Comment by long-xuan-nguyen** [comment]
1. A topology refresh did happen 
2. I dont enable debug log for topology view but the explaination makes sense
3. is there a log statement i can check for this?

null response was detected by our application (NullPointerException on ZCARD)

Is there an option to disable this behavior and workaround the issue while the root cause is being investigated?

**Comment by long-xuan-nguyen** [comment]
you mind link the code where the commands are re-dispatched? what class is responsible for detecting this scenario and re-send the commands in the buffer?

**Comment by long-xuan-nguyen** [comment]
autoReconnect is the option to disable sending commands on reconnection/switching to a different node? what is the effect of turning it off?

**Comment by long-xuan-nguyen** [comment]
just a note that some of the application servers recover just fine. we have 52 hosts and only a few of them have this issue

**Comment by mp911de** [comment]
`autoReconnect` is mainly responsible for automatic reconnection and whether to reject commands if the connection is broken.

You need to look into two places:

* [PooledClusterConnectionProvider:511](https://github.com/mp911de/lettuce/blob/4.1.x/src/main/java/com/lambdaworks/redis/cluster/PooledClusterConnectionProvider.java#L511)
* [ClusterNodeCommandHandler:74](https://github.com/mp911de/lettuce/blob/4.1.x/src/main/java/com/lambdaworks/redis/cluster/ClusterNodeCommandHandler.java#L74)

I figured out some bugs after having a look but I'm not sure that these bugs are necessarily responsible for the mentioned behavior:

* The code does not check whether the commands are canceled. That's more an optimization.
* Commands get completed twice in exception cases

**Comment by long-xuan-nguyen** [comment]
do you need any other log to help debugging the issue? i have not been able to reproduce the issue yet

Any pointer on how to reproduce?

**Comment by long-xuan-nguyen** [comment]
this is the log indicating topology refresh detecting the node is down; ie connection time out exception

03 Jun 2016 02:13:33,104 [DEBUG]  (lettuce-eventExecutorLoop-1-3) com.lambdaworks.redis.cluster.ClusterTopologyRefresh: Unable to connect to /10.88.161.176:6379
com.lambdaworks.redis.RedisConnectionException: Unable to connect to /10.88.161.176:6379
	at com.lambdaworks.redis.AbstractRedisClient.initializeChannel(AbstractRedisClient.java:250)
	at com.lambdaworks.redis.AbstractRedisClient.connectAsyncImpl(AbstractRedisClient.java:128)
	at com.lambdaworks.redis.cluster.RedisClusterClient.connectToNode(RedisClusterClient.java:346)
	at com.lambdaworks.redis.cluster.RedisClusterClient.connectToNode(RedisClusterClient.java:311)
	at com.lambdaworks.redis.cluster.ClusterTopologyRefresh.getConnections(ClusterTopologyRefresh.java:249)
	at com.lambdaworks.redis.cluster.ClusterTopologyRefresh.loadViews(ClusterTopologyRefresh.java:141)
	at com.lambdaworks.redis.cluster.RedisClusterClient$ClusterTopologyRefreshTask.run(RedisClusterClient.java:645)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:126)
	at io.netty.util.concurrent.DefaultEventExecutor.run(DefaultEventExecutor.java:56)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:742)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:145)
	at java.lang.Thread.run(Thread.java:745)
Caused by: io.netty.channel.ConnectTimeoutException: connection timed out: /10.88.161.176:6379
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe$1.run(AbstractNioChannel.java:284)
	at io.netty.util.concurrent.PromiseTask$RunnableAdapter.call(PromiseTask.java:38)
	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:120)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:339)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:373)

**Comment by long-xuan-nguyen** [comment]
the log does show once this exception occurs, the bad host is removed from topology

1 minute later, it is added back to the topology

**Comment by long-xuan-nguyen** [comment]
A couple of questions to help me understand the code better:

- is ClusterNodeCommandHandler responsible only for "cluster nodes" redis command? If so, why would Lettuce write commands like ZCARD into the underlying channel? I looked at RedisClusterClient and ClusterNodeCommandHandler is used only for topology refresh, it seems
- is CommandHandler meant to use as a singleton for all connections across all master/slave nodes?

**Comment by long-xuan-nguyen** [comment]
a couple more things i notice:

- during the rewrite of commands into S2, the response received is out of order; ie. an HGET result is returned to an EVALSHA call; unfortunately i can't show the log because it contains confidential information. This happens once. After that, all calls to S2 seem to return null
- an IllegalStateException in RSM is seen; this happens right before the above-mentioned out-of-order issue

03 Jun 2016 02:13:33,214 Caused by: java.lang.IllegalStateException
	at com.lambdaworks.redis.output.CommandOutput.set(CommandOutput.java:63)
	at com.lambdaworks.redis.protocol.RedisStateMachine.lambda$safeSet$0(RedisStateMachine.java:178)
	at com.lambdaworks.redis.protocol.RedisStateMachine.safeSet(RedisStateMachine.java:195)
	at com.lambdaworks.redis.protocol.RedisStateMachine.safeSet(RedisStateMachine.java:178)
	at com.lambdaworks.redis.protocol.RedisStateMachine.decode(RedisStateMachine.java:118)
	at com.lambdaworks.redis.protocol.CommandHandler.decode(CommandHandler.java:178)
	at com.lambdaworks.redis.protocol.CommandHandler.channelRead(CommandHandler.java:156)
	at io.netty.channel.ChannelHandlerInvokerUtil.invokeChannelReadNow(ChannelHandlerInvokerUtil.java:83)

**Comment by mp911de** [comment]
Thanks for your support. Each connection to a Redis server is composed of a `RedisChannelWriter` using the default implementation is `CommandHandler`. Every connection uses its own handler instances (see [here](https://github.com/mp911de/lettuce/blob/283801c6dfc784db677f0eb3714b2a5ab99efc56/src/main/java/com/lambdaworks/redis/ConnectionBuilder.java#L48-L68).

The default cluster connection uses `ClusterDistributionChannelWriter` as `RedisChannelWriter` to perform the routing and connections to Redis Cluster nodes use `ClusterNodeCommandHandler` to provide additional functionality.


Your only option to disable would be creating a subclass of `RedisClusterClient` in the `com.lambdaworks.redis.cluster` package and override `connectToNode(RedisCodec<K, V> codec, String nodeId, RedisChannelWriter<K, V> clusterWriter, Supplier<SocketAddress> socketAddressSupplier)`. You'd need to replace `clusterWriter` with `null`. This will cause commands to time out.

I'm not convinced the re-dispatching is the cause, I rather think that the client runs into some race-condition/synchronization issue.

**Comment by long-xuan-nguyen** [comment]
I think there is a race condition issue as well. As  a stop-gap, would you recommend modifying ClusterNodeCommandHandler and not re-write commands?

what log can i enable/put in to help troubleshooting if this issue comes up again?

I am still unable to reproduce this issue. Have you had any luck?

**Comment by mp911de** [comment]
Here are the steps how you can reproduce the issue with a leaving node and queued commands:

1. Have a master and a slave node, make sure to have periodic topology refreshes disabled and auto-reconnect enabled
2. Establish a connection to the master and set `ReadFrom.SLAVE`
3. Make sure the slave connection is initialized by issuing a `GET` command
4. Now come the tricky parts
5. Obtain the `StatefulRedisConnection` by the IP and port from `StatefulRedisClusterConnection`
6. Retrieve `ConnectionWatchdog` from the connection. See [here](https://github.com/mp911de/lettuce/blob/9278b4aa2a478fb1e7b6a3afc93272d301dac34a/src/test/java/com/lambdaworks/Connections.java#L16-L16) for how it's done. Now suspend auto-reconnect with `ConnectionWatchdog.setReconnectSuspended(false)`
7. Get `Partitions` from the `RedisClusterClient` and remove the slave node from `Partitions`, call `Partitions.updateCache()` afterwards
9. Disconnect the connection by closing the `Channel` with `Channel.close().sync()`. Your slave connection is now in a disconnected state.
10. Now fire a couple of commands reading commands. You should use different commands to hopefully trigger the out-of-order state.
11. Tell the client to apply the changed `Partitions` by calling `RedisClusterClient.updatePartitionsInConnections()`. That method is protected so either use reflection or subclass `RedisClusterClient`. This should be the final step.

I think you will need to invoke Steps 9 and 10 concurrently to simulate concurrency during the client usage. I expect no issues when calling Steps 9 and 10 sequentially, but the bug should get visible by adding some concurrency here. `ThreadPoolExecutor` might be the right approach here.

**Comment by long-xuan-nguyen** [comment]
Have you been able to reproduce the issue with these steps?

**Comment by mp911de** [comment]
Haven't tried yet, I'm currently busy with other issues.

**Comment by long-xuan-nguyen** [comment]
I tried with this

ClusterClientOptions options = new ClusterClientOptions.Builder()
                .closeStaleConnections(true)
                .refreshClusterView(false)
                .refreshPeriod(1, TimeUnit.SECONDS)
                .validateClusterNodeMembership(true)
                .pingBeforeActivateConnection(false)
                .autoReconnect(true)
                .build();

        RedisClusterClient clusterClient = RedisClusterClient.create(RedisURI.create("localhost", 5555));
        clusterClient.setOptions(options);

        StatefulRedisClusterConnection<String, String> connection = clusterClient.connect();
        connection.setReadFrom(ReadFrom.SLAVE);
        RedisAdvancedClusterCommands<String, String> client = connection.sync();


        System.out.println(client.zcard("key"));

        StatefulRedisConnection<String, String> a = connection.getConnection("127.0.0.1", 5555);

        ConnectionWatchdog connectionWatchdog = getConnectionWatchdog(a);
        connectionWatchdog.setReconnectSuspended(false);

        System.out.println(clusterClient.getPartitions());

        clusterClient.getPartitions().remove(clusterClient.getPartitions().getPartition(0));

        clusterClient.getPartitions().updateCache();

        System.out.println(clusterClient.getPartitions());

        getChannel(a).close().sync();

        ExecutorService executorService = Executors.newFixedThreadPool(500);

        for (int i = 0; i < 100000; i++) {
            final int j = i;
            executorService.submit(() -> {
               long number = client.zcard("key" + j);
                System.out.println(number);
                clusterClient.updatePartitionsInConnections();
                client.hget("key" + j, "hi");
                System.out.println("done");
            });
        }

        System.out.println("Waiting ...");

        executorService.awaitTermination(10, TimeUnit.DAYS);


However, once the slave node is disconnected, all the commands seem to be buffered. I was expecting ZCARD and HGET to be out of order and thus, NPE when casting ZCARD result to long.

Is there anything i am missing here?

**Comment by long-xuan-nguyen** [comment]
Regarding your comment to subclass RedisClusterClient and override connectToNode() to disable command re-dispatching, would it be sufficient to remove this block in ClusterNodeCommandHandler?

if (clusterChannelWriter != null) {
            if (isAutoReconnect() && !CHANNEL_OPEN_STATES.contains(getState())) {
                Collection<RedisCommand<K, V, ?>> commands = shiftCommands(queue);
                for (RedisCommand<K, V, ?> queuedCommand : commands) {
                    try {
                        clusterChannelWriter.write(queuedCommand);
                    } catch (RedisException e) {
                        queuedCommand.completeExceptionally(e);
                        queuedCommand.complete();
                    }
                }
            }

            Collection<RedisCommand<K, V, ?>> commands = shiftCommands(commandBuffer);
            for (RedisCommand<K, V, ?> queuedCommand : commands) {
                try {
                    clusterChannelWriter.write(queuedCommand);
                } catch (RedisException e) {
                    queuedCommand.completeExceptionally(e);
                }
            }
        }

I know it is not extensible but would this be a quick hack to avoid command re-dispatching which might possibly be the source of the race condition?

**Comment by long-xuan-nguyen** [comment]
and also, i can try removing command.complete() in ClusterNodeCommandHandler, which might make the client thinks the command is successful instead of treating it exceptionally. is that a possibility?

**Comment by mp911de** [comment]
Quick hacks won't happen. I was able to reproduce the out-of-order issue. The reason is canceled commands.

If you take a look [here](https://github.com/mp911de/lettuce/blob/9c0e165de32ebb6bdf9af49974e3ac53e1b06dfc/src/main/java/com/lambdaworks/redis/protocol/CommandHandler.java#L360) you'll notice that `queueCommand` performs a `isCanceled()` check. Commands are not added to the queue, but commands are written to the channel. That causes subsequent commands to be out of sync with the queue state as the queue contains only the new, healthy command but the response buffer contains the response of the previously sent command.


Regarding your code, it's always better to create Gists as they don't pollute the comments that much.

You need to move your lines

```
clusterClient.getPartitions().remove(clusterClient.getPartitions().getPartition(0));
clusterClient.getPartitions().updateCache();
```

after the `ThreadPool` code. 

Not sure `clusterClient.getPartitions().getPartition(0)` always hits the correct partition.

You also should invoke `clusterClient.updatePartitionsInConnections()` after the commands are submitted to the queue. You should reduce the excessive command count from `100000` to `20` or `50` and set a reasonable command timeout `clusterConnection.setTimeout(5, TimeUnit.SECONDS);`

Once you invoke synchronously a command on your connection after the `System.out.println("Waiting ...");` line, you should be able to reproduce the issue.

**Comment by long-xuan-nguyen** [comment]
Is the fix to do the same check for cancelled commands when re-dispatching them? Would it make more sense to have this check within CommandHandler?

**Comment by long-xuan-nguyen** [comment]
In the meantime, i've modified CommandHandler.java to not do ctx.write() if command is cancelled and it seems to resolve the issue.

Does that sound like a reasonable fix to you? 

When will you be fixing this in 4.2 and/or other branches?

**Comment by mp911de** [comment]
Writing/queueing only active commands is the right approach. I'm fixing this issue right now, you can have a look on the changeset.

**Comment by mp911de** [comment]
Thanks a lot for reporting and digging into the issue. It's great to receive such support. The issue is fixed and I plan to release lettuce 4.2.0 today.
